### PR TITLE
DropDaemon refactoring

### PR DESCRIPTION
**Pull request by olkononenko**
[AB#1552016](https://dev.azure.com/mseng/708e929f-6bd5-415a-8daf-25b1dac08dd8/_workitems/edit/1552016)

### CODE REVIEW

File: Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+
+        public static readonly BoolOption Verbose = RegisterDaemonConfigOption(new BoolOption("verbose")
+        {
+            ShortName = "v",
+            HelpText = "Verbose logging",
+            IsRequired = false,
+            DefaultValue = false,
+        });
+
+        public static readonly StrOption DropServiceConfigFile = RegisterDaemonConfigOption(new StrOption("dropServiceConfigFile")
```
Comment (narasamdya) [inline Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs]: >DropServiceConfigFile [](start = 41, length = 21)

Is this not specify for Drop?

File: Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+        {
+            options.Add(option);
+            return option;
+        }
+
+        protected static T RegisterDaemonConfigOption<T>(T option) where T : Option => RegisterOption(DaemonConfigOptions, option);
+
+        /// <remarks>
+        /// The <see cref="DaemonConfigOptions"/> options are added to every command.
+        /// A non-mandatory string option "name" is added as well, which drop operation
```
Comment (narasamdya) [inline Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs]: >drop operation [](start = 73, length = 14)

Seems to be drop specific.

File: Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+            if (!Guid.TryParse("43b71382-88db-5427-89d5-0b46476f8ef4", out Guid guid))
+            {
+                Error("Could not parse guid; exiting");
+                return -1;
+            }
+
+            Console.WriteLine("Listening for cloud build events");
+
+            // Create an unique session
+            string sessionName = "DropDaemon ETW Session";
```
Comment (narasamdya) [inline Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs]: >DropDaemon ETW Session [](start = 34, length = 22)

Do we need different ETW sessions for drop and symbol?

File: Public/Src/Tools/DropDaemon/DropDaemon.cs [hunk 4]
```diff
@@ -0,0 +1,875 @@
...
+                {
+                    dropEvent.DropUrl = (await m_dropClientTask).DropUrl;
+                }
+
+                // send event
+                m_etwLogger.Log(dropEvent);
+            }
+        }
+
+        #region Config options and commands
```
Comment (narasamdya) [inline Public/Src/Tools/DropDaemon/DropDaemon.cs]: >#region Config options and commands [](start = 8, length = 35)

Is it worth splitting into different files?

File: Public/Src/Tools/ServicePipDaemon/Statics.cs [hunk 5]
```diff
@@ -5,33 +5,34 @@
...
-        /// <summary>
-        ///     String format with <see cref="CultureInfo.InvariantCulture"/>.
-        /// </summary>
-        public static string Inv(string format, params object[] args)
-        {
-            Contract.Requires(format != null);
-
-            return string.Format(CultureInfo.InvariantCulture, format, args);
-        }
+        ///// <summary>
```
Comment (almiliMSFT) [inline Public/Src/Tools/ServicePipDaemon/Statics.cs]: remove commented out code completely?

File: Public/Src/Tools/ServicePipDaemon/Tool.ServicePipDaemon.dsc [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+import * as BuildXLSdk from "Sdk.BuildXL";
+import * as Managed from "Sdk.Managed";
+import * as Deployment from "Sdk.Deployment";
+import { NetFx } from "Sdk.BuildXL";
+
+namespace ServicePipDaemon {
+
+    @@public
+    export const dll = !BuildXLSdk.isDropToolingEnabled ? undefined : BuildXLSdk.library({
```
Comment (almiliMSFT) [inline Public/Src/Tools/ServicePipDaemon/Tool.ServicePipDaemon.dsc]: nit: Consider renaming `isDropToolingEnabled`
```suggestion
    export const dll = !BuildXLSdk.isArtifactServicesToolingEnabled ? undefined : BuildXLSdk.library({
```

File: Public/Src/Tools/ServicePipDaemon/Tool.ServicePipDaemon.dsc [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        assemblyName: "ServicePipDaemon",
+        //rootNamespace: "Tool.ServicePipDaemon",
```
Comment (almiliMSFT) [inline Public/Src/Tools/ServicePipDaemon/Tool.ServicePipDaemon.dsc]:
```suggestion
        rootNamespace: "Tool.ServicePipDaemon",
```

File: Public/Src/Tools/ServicePipDaemon/Tool.ServicePipDaemon.dsc [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        skipDocumentationGeneration: true,
```
Comment (almiliMSFT) [inline Public/Src/Tools/ServicePipDaemon/Tool.ServicePipDaemon.dsc]: is there still a reason to skip generating documentation?

File: Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+            }
+
+            // BuildXL.Tracing.ETWLogger guid
+            if (!Guid.TryParse("43b71382-88db-5427-89d5-0b46476f8ef4", out Guid guid))
+            {
+                Error("Could not parse guid; exiting");
+                return -1;
+            }
+
+            Console.WriteLine("Listening for cloud build events");
```
Comment (almiliMSFT) [inline Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs]: Can this go through the logger instead of directly to Console?

Comment (almiliMSFT) [reply Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs]: never mind, this is just a command-line program for testing

File: Public/Src/Tools/DropDaemon/DropDaemon.cs [hunk 10]
```diff
@@ -0,0 +1,875 @@
...
+            {
+                // log stats
+                m_logger.Info("Statistics: ");
+                m_logger.Info(string.Join(Environment.NewLine, stats.Select(s => s.Key + " = " + s.Value)));
+
+                stats.Add(nameof(DropConfig) + nameof(DropConfig.MaxParallelUploads), DropConfig.MaxParallelUploads);
+                stats.Add(nameof(DropConfig) + nameof(DropConfig.NagleTime), (long)DropConfig.NagleTime.TotalMilliseconds);
+                stats.Add(nameof(DropConfig) + nameof(DropConfig.BatchSize), DropConfig.BatchSize);
+                stats.Add(nameof(DropConfig) + nameof(DropConfig.EnableChunkDedup), DropConfig.EnableChunkDedup ? 1 : 0);
+                stats.Add("DaemonConfig" + nameof(Config.MaxConcurrentClients), Config.MaxConcurrentClients);
```
Comment (almiliMSFT) [inline Public/Src/Tools/DropDaemon/DropDaemon.cs]: scope creep:
```suggestion
                stats.Add(nameof(DaemonConfig) + nameof(Config.MaxConcurrentClients), Config.MaxConcurrentClients);
```

File: Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs [hunk 11]
```diff
@@ -0,0 +1,538 @@
...
+            ConfiguredCommand conf;
+            using (m_counters.StartStopwatch(DaemonCounter.ParseArgsDuration))
+            {
+                conf = ParseArgs(cmdLine, m_parser);
+            }
+
+            IIpcResult result;
+            using (m_counters.StartStopwatch(DaemonCounter.ServerActionDuration))
+            {
+                 result = await conf.Command.ServerAction(conf, this);
```
Comment (almiliMSFT) [inline Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs]: Just a comment for future use: it would be very helpful if we stick this `ServerActionDuration` time into `result`.

File: Public/Src/Tools/DropDaemon/DropDaemon.cs (hunk 4 above)
Comment (olkononenko) [reply Public/Src/Tools/DropDaemon/DropDaemon.cs]: Due to how things (options/commands) are interconnected, options/commands should stay in one place. Moving them into a separate (sub)class won't work either (well, it will look even uglier).

---
In reply to: [298344104](https://github.com/microsoft/BuildXL/pull/528#discussion_r298344104) [](ancestors = 298344104)
